src/store.rs

```diff
@@ -0,0 +1,196 @@
+            root,
+            inputs_to_content,
+        })
+    }
+
+    pub fn for_job(&self, job: &Job) -> Option<PathBuf> {
+        self.inputs_to_content
+            .get(&job.id)
+            .map(|path| self.root.join(path))
+    }
+
+    pub fn store_from_workspace(&mut self, job: &Job, workspace: Workspace) -> Result<()> {
+        let mut hasher = blake3::Hasher::new();
+        let temp = tempfile::Builder::new()
+            .prefix(&format!("rbt-job-{}", job.id))
+            .tempdir()
+            .context("couldn't create temporary directory for hashing")?;
+
+        let mut output_dirs: HashSet<PathBuf> = HashSet::new();
+
+        for output in job.outputs.iter().sorted() {
+            let source = PathBuf::from(output.as_str());
+
+            ///////////////////////////////////////////////////
+            // Step 1: Validate that the path we got is safe //
+            ///////////////////////////////////////////////////
+            for component in source.components() {
+                match component {
+                    Component::Prefix(_) | Component::RootDir =>  anyhow::bail!(
+                        "Got `{}` as an output, but absolute paths are not allowed as outputs. Remove the absolute prefix to fix this!",
+                        source.display(),
+                    ),
+
+                    Component::ParentDir => anyhow::bail!(
+                        "Got `{}` as an output, but relative paths containing `..` are not allowed as inputs. Remove the `..` to fix this!",
+                        source.display(),
+                    ),
+
+                    Component::CurDir | Component::Normal(_) => (),
+                };
+            }
+
+            ////////////////////////////////////////////////////////////////
+            // Step 2: add the file content to the content-addressed hash //
+            ////////////////////////////////////////////////////////////////
+            hasher.update(output.as_bytes());
+
+            let mut file = File::open(&workspace.join(output.as_str())).with_context(|| {
+                format!(
+                    "couldn't open `{}` for hashing. Did the build produce it?",
+                    output
+                )
+            })?;
+
+            // TODO: docs for Blake3 say that a 16 KiB buffer is the most
+            // efficient (for SIMD reasons), but `std::io::copy` uses an 8KiB
+            // buffer. Gonna have to do this by hand at some point to take
+            // advantage of the algorithm's designed speed.
+            std::io::copy(&mut file, &mut hasher).with_context(|| {
+                format!("could not read `{}` to calculate hash", source.display())
+            })?;
+
+            ///////////////////////////////////////////////
+            // Step 3: make sure any parent paths exist  //
+            ///////////////////////////////////////////////
+            let mut ancestors: Vec<&Path> = source.ancestors().skip(1).collect();
```

Comment:
ancestors for `a/b/c" will look like `["a/b/c", "a/b", "a", ""]` ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=4c1ba455ad1061470830c74e1c5bd7c7)) From those:

- `skip(1)` because we don't want "a/b/c" because that's the full path—if we create a directory there, we'll conflict with the output.
- `pop()` because we also don't want the path's final base (in this case always "" since we've determined the path is relative already.)

Reply:
ahh, different sides of the iterator.